lib/questionlib.php

```diff
@@ -2594,3 +2076,726 @@ function get_next_version(int $questionbankentryid): ?int {
+     * @return moodle_url the URL.
+     * @deprecated since Moodle 4.0
+     * @see qbank_previewquestion\previewquestion_helper::question_preview_url()
+     */
+    function question_preview_url($questionid, $preferredbehaviour = null,
+                                  $maxmark = null, $displayoptions = null, $variant = null, $context = null) {
+        debugging('Function question_preview_url() has been deprecated and moved to qbank_previewquestion plugin,
+     Please use qbank_previewquestion\previewquestion_helper::question_preview_url() instead.', DEBUG_DEVELOPER);
+
+        return \qbank_previewquestion\previewquestion_helper::question_preview_url($questionid, $preferredbehaviour = null,
+            $maxmark = null, $displayoptions = null, $variant = null, $context = null);
+    }
+
+    /**
+     * @return array that can be passed as $params to the {@link popup_action} constructor.
+     * @deprecated since Moodle 4.0
+     * @see qbank_previewquestion\previewquestion_helper::question_preview_popup_params()
+     */
+    function question_preview_popup_params() {
+        debugging('Function question_preview_popup_params() has been deprecated and moved to qbank_previewquestion plugin,
+    Please use qbank_previewquestion\previewquestion_helper::question_preview_popup_params() instead.', DEBUG_DEVELOPER);
+
+        return \qbank_previewquestion\previewquestion_helper::question_preview_popup_params();
+    }
+
+    /**
+     * Creates a stamp that uniquely identifies this version of the question
+     *
+     * In future we want this to use a hash of the question data to guarantee that
+     * identical versions have the same version stamp.
+     *
+     * @param object $question
+     * @return string A unique version stamp
+     * @deprecated since Moodle 4.0
+     */
+    function question_hash($question) {
+        debugging('Function question_hash() has been deprecated without replacement.', DEBUG_DEVELOPER);
+        return make_unique_id_code();
+    }
+
+    /**
+     * Private method, only for the use of add_indented_names().
+     *
+     * Recursively adds an indentedname field to each category, starting with the category
+     * with id $id, and dealing with that category and all its children, and
+     * return a new array, with those categories in the right order.
+     *
+     * @param array $categories an array of categories which has had childids
+     *          fields added by flatten_category_tree(). Passed by reference for
+     *          performance only. It is not modfied.
+     * @param int $id the category to start the indenting process from.
+     * @param int $depth the indent depth. Used in recursive calls.
+     * @param int $nochildrenof
+     * @return array a new array of categories, in the right order for the tree.
+     * @deprecated since Moodle 4.0 MDL-71585
+     * @see qbank_managecategories\helper
+     */
+    function flatten_category_tree(&$categories, $id, $depth = 0, $nochildrenof = -1) {
```

Comment:
remove the deprecated code and point to the new function

Reply:
done, changed to public in managecategories.